### PR TITLE
feat(ui): dropdown component add

### LINKED ISSUE #523: Add dropdown component

### PR DESCRIPTION
**Pull request by 32penkin**
### Please read and mark the following checklist before creating a pull request:

 - [x] I read and followed the [CONTRIBUTING.md](https://github.com/akveo/react-native-ui-kitten/blob/master/CONTRIBUTING.md) guide.
 - [x] I read and followed the [New Feature Checklist](https://github.com/akveo/react-native-ui-kitten/blob/master/DEV_DOCS.md) guide.

* Closes #523

### CODE REVIEW

File: src/framework/ui/dropdown/droppdownItem.component.tsx [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    return {
+      container: containerStyles,
+      text: {
+        color: textColor,
+        fontSize: textFontSize,
+        fontWeight: textFontWeight,
+        lineHeight: textLineHeight,
+        marginHorizontal: textMarginHorizontal,
+      },
+      containerSelected: {
```
Comment (artyorsh) [inline src/framework/ui/dropdown/droppdownItem.component.tsx]: Looks like this can be implemented with Eva states

**Comment by artyorsh** [comment]
Let's first merge this one #542 and update the component

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+type IconElement = React.ReactElement<ImageProps>;
+type MenuElement = React.ReactElement<DropdownMenuProps>;
+type ControlElement = React.ReactElement<TouchableOpacityProps>;
+type IconProp = (style: ImageStyle, visible: boolean) => IconElement;
+export type SelectedOption = Array<DropdownItemType> | DropdownItemType;
+
+const MEASURED_CONTROL_TAG: string = 'Control';
+
+interface ComponentProps {
+  items: DropdownItemType[];
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Done

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+  public state: State = {
+    visible: false,
+    menuWidth: 0,
+  };
+
+  public componentWillMount(): void {
+    const { multiSelect, selectedOption } = this.props;
+
+    this.strategy = multiSelect ?
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Done

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+      new MultiSelectStrategy(selectedOption) : new SingleSelectStrategy(selectedOption);
+  }
+
+  private strategy: SelectionStrategy;
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Done

File: src/framework/ui/dropdown/selection.strategy.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  constructor(options: DropdownItemType | DropdownItemType[]) {
+    if (Array.isArray(options)) {
+      this.selectedOption = options;
+    }
+  }
+
+  public select(option: DropdownItemType): void {
+    const optionAlreadyExist: boolean = this.selectedOption
+      .some((item: DropdownItemType) => item === option);
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/selection.strategy.ts]: Done.

File: src/framework/ui/dropdown/selection.strategy.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+  public getPlaceholder(placeholder: string): string {
+    if (this.isSelectedOptionExist()) {
+      return this.selectedOption.map((item: DropdownItemType) => item.text).join(', ');
+    } else {
+      return placeholder;
+    }
+  }
+
+  public onSelect(): DropdownItemType[] {
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/selection.strategy.ts]: Done.

File: src/framework/ui/dropdown/selection.strategy.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    return this.selectedOption;
+  }
+
+  public isSelected(item: DropdownItemType): boolean {
+    return this.selectedOption.some((option: DropdownItemType) => option === item);
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/selection.strategy.ts]: Done.

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    const { control, icon, text } = this.getComponentStyle(themedStyle);
+
+    const iconElement: IconElement = this.renderIconElement(icon);
+    const textElement: TextElement = this.renderTextElement(text);
+
+    const measuringProps: MeasuringElementProps = { tag: MEASURED_CONTROL_TAG };
+
+    return (
+      <MeasureNode onResult={this.onControlMeasure}>
+        {[
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Will be in the separate PR

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+  MeasureResult,
+  MeasuringElementProps,
+} from '../popover/measure.component';
+import {
+  SelectionStrategy,
+  MultiSelectStrategy,
+  SingleSelectStrategy,
+} from './selection.strategy';
+
+type TextElement = React.ReactElement<TextProps>;
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Sure. Let's do it in the separate PR

File: src/framework/ui/dropdown/droppdownItem.component.tsx (hunk 1 above)
Comment (32penkin) [reply src/framework/ui/dropdown/droppdownItem.component.tsx]: Done

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        color: labelStyle.labelColor,
+        marginBottom: labelStyle.labelMarginBottom,
+      },
+    };
+  };
+
+  private renderLabelElement = (style: TextStyle): TextElement | null => {
+    const { label, labelStyle } = this.props;
+
+    return label ? (
```
[the remark this answers is not in the document]
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Sure, done

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 11]
```diff
@@ -51,22 +51,22 @@ type ControlElement = React.ReactElement<TouchableOpacityProps>;
...
   placeholderStyle?: StyleProp<TextStyle>;
   label?: string;
   labelStyle?: StyleProp<TextStyle>;
   controlStyle?: StyleProp<ViewStyle>;
   icon?: IconProp;
-  onSelect: (option: SelectedOption, event?: GestureResponderEvent) => void;
+  onSelect: (option: DropdownOption, event?: GestureResponderEvent) => void;
   size?: string;
   status?: string;
   appearance?: string;
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdown.component.tsx]: has declaration in `StyledComponentProps`

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 12]
```diff
@@ -115,8 +115,11 @@ class DropdownComponent extends React.Component<DropdownProps, State> {
     this.setState({ visible });
   };
 
-  private onPress = () => {
+  private onPress = (event: GestureResponderEvent) => {
     this.props.dispatch([]);
+    if (this.props.onPress) {
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdown.component.tsx]: Do we have any expected behavior when we pass `onPress` property?

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 13]
```diff
@@ -226,7 +229,7 @@ class DropdownComponent extends React.Component<DropdownProps, State> {
   };
 
   private renderMenuElement = (style: StyleType): DropdownMenuElement => {
-    const { appearance, selectedOption, ...restProps } = this.props;
+    const { appearance, selectedOption, testID, ...restProps } = this.props;
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdown.component.tsx]: I don't think this is a correct way to test components.
You can use any of [getAllBy](https://callstack.github.io/react-native-testing-library/docs/api-queries#getallby) queries to find a corresponding MenuElement

File: src/framework/ui/dropdown/dropdown.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+        onDropdownLongPress={onDropdownLongPress}
+      />,
+    );
+
+    fireEvent(application.getByTestId(dropdownTestId), 'pressIn');
+    fireEvent(application.getByTestId(dropdownTestId), 'pressOut');
+    fireEvent(application.getByTestId(dropdownTestId), 'longPress');
+    fireEvent.press(application.getAllByType(TouchableOpacity)[0]);
+
+    const { label, placeholder } = application.getByTestId(dropdownTestId).props;
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdown.spec.tsx]: This is also incorrect way of testing. Expected behavior is that passed `label` and `placeholder` values are finally rendered somewhere. For example, you can use [getByText](https://callstack.github.io/react-native-testing-library/docs/api-queries#bytext) query and expect `children` `toEqual` `label` for this purpose

Comment (artyorsh) [reply src/framework/ui/dropdown/dropdown.spec.tsx]: Let's also split this case into multiple cases for each property

File: src/framework/ui/dropdown/dropdownMenu.component.tsx [hunk 15]
```diff
@@ -68,6 +68,7 @@ export class DropdownMenu extends React.Component<DropdownMenuProps> {
     ) : (
       <DropdownItem
         {...info}
+        {...info.item}
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdownMenu.component.tsx]: Is it expected that a final object of properties will look like the one below?
```js
{
  item: {
    ...itemProps, // <-- props of the `item` 
  },
  ...itemProps, // <-- same props without nesting in `item`
  index,
  // ...
}
```

File: src/framework/ui/dropdown/droppdownItem.component.tsx [hunk 16]
```diff
@@ -31,6 +31,8 @@ type TextElement = React.ReactElement<TextProps>;
 type DefaultItemElement = React.ReactElement<TouchableOpacityProps>;
 type MultiSelectItemElement = React.ReactElement<ViewProps>;
 
+const itemTestId = (text: string): string => `@dropdown-item/${text}`;
```
Comment (artyorsh) [inline src/framework/ui/dropdown/droppdownItem.component.tsx]: There is a better way to do this (with `getAllBy*` query). Explained above

File: src/playground/src/ui/screen/dropdown/dropdown.container.tsx [hunk 17]
```diff
@@ -87,7 +88,7 @@ export class DropdownContainer extends React.Component<any, State> {
         <Dropdown
           style={styles.dropdown}
           label='Options'
-          placeholder='Select Option'
+          // placeholder='Select Option'
```
Comment (artyorsh) [inline src/playground/src/ui/screen/dropdown/dropdown.container.tsx]: I'd prefer to have it :)

File: src/framework/ui/dropdown/dropdown.component.tsx (hunk 12 above)
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: I don't really keep it in mind. Just for test coverage

File: src/framework/ui/dropdown/dropdown.component.tsx (hunk 13 above)
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Refactored

File: src/framework/ui/dropdown/dropdown.spec.tsx (hunk 14 above)
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.spec.tsx]: refactored

File: src/framework/ui/dropdown/droppdownItem.component.tsx (hunk 16 above)
Comment (32penkin) [reply src/framework/ui/dropdown/droppdownItem.component.tsx]: refactored

File: src/playground/src/ui/screen/dropdown/dropdown.container.tsx (hunk 17 above)
Comment (32penkin) [reply src/playground/src/ui/screen/dropdown/dropdown.container.tsx]: Will be in Showcase

File: src/framework/ui/dropdown/dropdown.component.tsx (hunk 11 above)
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: Removed

File: src/framework/ui/dropdown/dropdownMenu.component.tsx (hunk 15 above)
Comment (32penkin) [reply src/framework/ui/dropdown/dropdownMenu.component.tsx]: Refactored

**Comment by artyorsh** [comment]
Also, please add `DropdownElement` to exports

**Comment by Juanperezc** [comment]
Amazing, I wait for this

File: src/framework/ui/dropdown/dropdown.component.tsx [hunk 18]
```diff
@@ -0,0 +1,751 @@
...
+      this.renderMenuElement(style.menu),
+      isValidString(label) && this.renderLabelElement(style.label),
+      this.renderControlElement(),
+    ];
+  };
+
+  public render(): DropdownElement {
+    const { themedStyle, style } = this.props;
+    const { visible, menuWidth } = this.state;
+    const evaStyles: StyleType = this.getComponentStyle(themedStyle);
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdown.component.tsx]: let's use our common naming for this variable

File: src/framework/ui/dropdown/dropdownMenu.component.tsx [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+  private onSelect = (option: DropdownItemType, event?: GestureResponderEvent): void => {
+    this.props.onSelect(option, event);
+  };
+
+  private renderDefaultItem = (info: ListRenderItemInfo<DropdownItemType>): DefaultMenuItemElement => {
+    const { renderItem, multiSelect, strategy } = this.props;
+    const selected: boolean = strategy.isSelected(info.item);
+
+    return this.areThereSubItems(info.item) ? (
```
Comment (artyorsh) [inline src/framework/ui/dropdown/dropdownMenu.component.tsx]: can be structured better (e.g with render props)

File: src/framework/ui/dropdown/dropdown.component.tsx (hunk 18 above)
Comment (32penkin) [reply src/framework/ui/dropdown/dropdown.component.tsx]: done

**Pull request merged**